src/index.ts

```diff
@@ -0,0 +1,40 @@
+import ConstructedStyleSheet from './ConstructedStyleSheet';
+import {
+  attachAdoptedStyleSheetProperty,
+  getAssociatedLocation,
+} from './Location';
+import {closedShadowRootRegistry} from './shared';
+
+// @ts-expect-error: we need to replace the default CSSStyleSheet class with
+// the emulation.
+window.CSSStyleSheet = ConstructedStyleSheet;
+
+attachAdoptedStyleSheetProperty(Document);
+
+if ('ShadowRoot' in window) {
+  attachAdoptedStyleSheetProperty(ShadowRoot);
+
+  var proto = Element.prototype;
+  var attach = proto.attachShadow;
+
+  proto.attachShadow = function attachShadow(init) {
+    var root = attach.call(this, init);
+
+    if (init.mode === 'closed') {
+      closedShadowRootRegistry.set(this, root);
```

Comment:
Would it be more performant to save a reference to all shadow roots so there's less DOM traversal later?

Reply:
Unfortunately, no because we still need to understand if the element is connected. So the traversal should be done as it is now. It may allow avoid using the `element.shadowRoot || closedShadowRootRegistry.get(element)` expression but I think it would be more memory-friendly to keep this map only for closed shadow roots. Correct me if I'm wrong: it may be an unnecessary micro-optimization.